crypto/fipsmodule/rsa/rsa_impl.c

```diff
@@ -113,6 +108,27 @@ int rsa_check_public_key(const RSA *rsa) {
     OPENSSL_PUT_ERROR(RSA, RSA_R_KEY_SIZE_TOO_SMALL);
     return 0;
   }
+
+  return 1;
+}
+
+rsa_key_type_t detect_rsa_key_type(const RSA *rsa) {
+  if (rsa->d == NULL || BN_is_zero(rsa->d)) {
```

Comment:
I'm not sure this is OK. Technically, you don't need `d` for RSA-CRT, you just need the intermediate values to compute the CRT.

It at least looks like the default signing implementation wouldn't allow this though:
```
int rsa_default_private_transform(RSA *rsa, uint8_t *out, const uint8_t *in,
                                  size_t len) {
  if (rsa->n == NULL || rsa->d == NULL) {
    OPENSSL_PUT_ERROR(RSA, RSA_R_VALUE_MISSING);
    return 0;
  }
```

But who knows - maybe in the future this will not be the case and we will infer the wrong key type here.

We might want to rethink this function. It is also a behaviour change no? Since non-private keys before were only identified by looking at `p/q`.

Reply:
yeah, i agree that we should try to "future-proof" the key type detection logic to the best of our ability.

while there is the check on `d` there in the default private transform, i don't believe `d` is actually used when the [CRT params are present](https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/rsa/rsa_impl.c#L752) (i.e. function `mod_exp` is called, and i don't see any uses of `d` in [that function](https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/rsa/rsa_impl.c#L849-L930), only `d mod (p-1)`). when the CRT params are omitted, `d` [is used](https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/rsa/rsa_impl.c#L763). so, i could see a world in which we relax the NULL check on `d` at the top of `rsa_default_private_transform`.

how about these criteria for determining key type?

- PUBLIC: `e` is present and non-zero, and any of `p`, `q`, or `d` are absent or zero
- PRIVATE_STRIPPED: `d` is present and non-zero, and any of `p`, `q`, or `iqmp` are absent or zero, or `dmp1` or `dmq1` are absent
- PRIVATE_CRT: all of `p`, `q`, and `iqmp` are present and non-zero, and `dmp1` and `dmq1` are both present
- UNKNOWN: none of the above conditions met. we return an error immediately when we detect an unknown key type.

Reply:
`d` is not used in decryption because `d mod (p-1)` and `d mod (q-1)` are already stored as part of the key. So I think classifying a key as PRIVATE_CRT should be done when: `p` and `q` are present and `d` or `(d mod (p-1), d mod (q-1))` is present. Since the existing functions already have the check on `d` and with `d` one can compute `(d mod (p-1), d mod (q-1))`, I would rather leave it in the new function as well.

Reply:
ah, that makes sense. thanks dusan, i'll implement that and update the tests accordingly.

Reply:
The private key operation implementations support not having CRT values in the RSA object. So, not sure this would be a valid classification.